algorithms.py

```diff
@@ -15,58 +17,138 @@
 X = np.random.normal(size = (m, n))
 X = X - np.mean(X, axis=1, keepdims=True)
 
+W1 = np.random.normal(size = (k, m))
+W2 = np.random.normal(size = (m, k))
+
+# auxiliary function that computes the distance between two SVDs
+def compute_dist(W2,k,u_svd):
+    dist = 0;
+    u, s, _ = np.linalg.svd(W2, full_matrices = False)
+    for col in range(k):
+        dist += np.min([np.linalg.norm(u_svd[:,col]-u[:,col]), np.linalg.norm(-u_svd[:,col]-u[:,col])])
+    return dist
+
 # SVD with dgesdd (divide and conquer)
-start = time.time()
-u, s, _ = np.linalg.svd(X, full_matrices = False)
-end = time.time()
-t = end - start
+def svd():
+    start = time.time()
+    u, s, _ = np.linalg.svd(X, full_matrices = False)
+    end = time.time()
+    t = end - start
+    return (u,s,t)
 
 # Randomized SVD
-start = time.time()
-u1, s1, _ = randomized_svd(X, 2)
-end = time.time()
-t1 = end - start
+def randomized_svd_wrapper():
+    start = time.time()
+    u, s, _ = randomized_svd(X, 2)
+    end = time.time()
+    t = end - start
+    return (u,s,t)
 
-# LAE-PCA
-start = time.time()
-W1 = np.random.normal(size = (k, m))
-W2 = np.random.normal(size = (m, k))
-XXt = X @ X.T
-i = 0
-while np.linalg.norm(W1 - W2.T) > eps:
-    W1 -= alpha * ((W2.T @ (W2 @ W1 - I)) @ XXt + lamb * W1)
-    W2 -= alpha * (((W2 @ W1 - I) @ XXt) @ W1.T + lamb * W2)
-    i += 1
-u2, s2, _ = np.linalg.svd(W2, full_matrices = False)
-s2 = np.sqrt(lamb / (1 - s2**2))
-end = time.time()
-t2 = end - start
+# LAE-PCA using gradient descent
+def LAE_PCA_GD(W1,W2,u_svd = None):
+    W1 = W1.copy()
+    W2 = W2.copy()
+    start = time.time()
+    XXt = X @ X.T
+    i = 0
+    dist = []
+    while np.linalg.norm(W1 - W2.T) > eps:
+        W1 -= alpha * ((W2.T @ (W2 @ W1 - I)) @ XXt + lamb * W1)
+        W2 -= alpha * (((W2 @ W1 - I) @ XXt) @ W1.T + lamb * W2)
+        
+        if u_svd is not None:
+            dist.append( compute_dist(W2,k,u_svd) )
+        i += 1
+        
+    u, s, _ = np.linalg.svd(W2, full_matrices = False)
+    s = np.sqrt(lamb / (1 - s**2))
+    end = time.time()
+    t = end - start
+    return (u,s,t,i,dist)
 
 # Regularized Oja's Rule
-start = time.time()
-XXt = X @ X.T
-W2 = np.random.normal(size = (m, k))
-diff = np.inf
-j = 0
-while diff > eps:
-    update = alpha * (((W2 @ W2.T - I) @ XXt) @ W2 + lamb * W2)
-    W2 -= update
-    diff = np.linalg.norm(update)
-    j += 1
-u3, s3, _ = np.linalg.svd(W2, full_matrices = False)
-s3 = np.sqrt(lamb / (1 - s3**2))
-end = time.time()
-t3 = end - start
-
-print('SVD ({:0.5f} secs)'.format(t))
-print(s[0:k])
-print(u[:, 0:k])
-print('Randomized SVD ({:0.5f} secs)'.format(t1))
-print(s1[0:k])
-print(u1[:, 0:k])
-print('LAE-PCA  ({:0.5f} secs, {} iterations)'.format(t2, i))
-print(s2[0:k])
-print(u2[:, 0:k])
-print('Regularized Oja\'s rule ({:0.5f} secs, {} iterations)'.format(t3, j))
-print(s3)
-print(u3)
+def LAE_PCA_Oja_GD(W2, u_svd = None):
```

Comment:
no need for _GD here, since it's the only option. change the plot label as well

Reply:
I do not see why GD is the only applicable algorithm.

Reply:
I understood your comment to be that exact could not be applied efficiently (or at least simply) to Oja.  But I'm fine leaving GD

Reply:
exact is probably impossible, but the world is bigger than GD and exact